openhands/core/message.py

```diff
@@ -48,34 +49,48 @@ def serialize_model(self):
 
 
 class Message(BaseModel):
-    role: Literal['user', 'system', 'assistant']
-    content: list[TextContent | ImageContent] = Field(default=list)
+    role: Literal['user', 'system', 'assistant', 'tool']
+    content: list[TextContent | ImageContent] = Field(default_factory=list)
     cache_enabled: bool = False
     vision_enabled: bool = False
+    # function calling
+    # - tool calls (from LLM)
+    tool_calls: list[ChatCompletionMessageToolCall] | None = None
+    # - tool execution result (to LLM)
+    tool_call_id: str | None = None
+    name: str | None = None  # name of the tool
 
     @property
     def contains_image(self) -> bool:
         return any(isinstance(content, ImageContent) for content in self.content)
 
     @model_serializer
     def serialize_model(self) -> dict:
-        content: list[dict] | str
-        # two kinds of serializer:
-        # 1. vision serializer: when prompt caching or vision is enabled
-        # 2. single text serializer: for other cases
-        # remove this when liteLLM or providers support this format translation
-        if self.cache_enabled or self.vision_enabled:
-            # when prompt caching or vision is enabled, use vision serializer
-            content = []
-            for item in self.content:
-                if isinstance(item, TextContent):
-                    content.append(item.model_dump())
-                elif isinstance(item, ImageContent):
-                    content.extend(item.model_dump())
-        else:
-            # for other cases, concatenate all text content
-            # into a single string per message
-            content = '\n'.join(
-                item.text for item in self.content if isinstance(item, TextContent)
-            )
-        return {'content': content, 'role': self.role}
+        content: list[dict] = []
+        role_tool_with_prompt_caching = False
+        for item in self.content:
+            d = item.model_dump()
+            # We have to remove cache_prompt for tool content and move it up to the message level
+            # See discussion here for details: https://github.com/BerriAI/litellm/issues/6422#issuecomment-2438765472
```

Comment:
@xingyaoww  I saw the discussion, but it looks like our code completely removed the single-string serializer? It wasn't just for this, it was needed for LLMs or providers that don't support list of strings/images at all, they just need a single string per message (sadly). I think it's the old version of openAI API, and not everyone has adapted to the new one.

That said, this serializer was intended as temporary. If they adapt or LiteLLM does handle it completely, then we will not need to do this. I'll test to see where we are now.